src/bootstrap/src/core/build_steps/perf.rs

```diff
@@ -0,0 +1,45 @@
+use std::process::Command;
+
+use crate::core::build_steps::compile::{Std, Sysroot};
+use crate::core::build_steps::tool::RustcPerf;
+use crate::core::builder::Builder;
+use crate::core::config::DebuginfoLevel;
+
+/// Performs profiling using `rustc-perf` on a built version of the compiler.
+pub fn perf(builder: &Builder<'_>) {
+    let collector = builder.ensure(RustcPerf {
+        compiler: builder.compiler(0, builder.config.build),
+        target: builder.config.build,
+    });
+
+    if builder.build.config.rust_debuginfo_level_rustc == DebuginfoLevel::None {
+        builder.info(r#"WARNING: You are compiling rustc without debuginfo, this will make profiling less useful.
+Consider setting `rust.debuginfo-level = 1` in `config.toml`."#);
+    }
+
+    let compiler = builder.compiler(builder.top_stage, builder.config.build);
+    builder.ensure(Std::new(compiler, builder.config.build));
+    let sysroot = builder.ensure(Sysroot::new(compiler));
+    let rustc = sysroot.join("bin/rustc");
+
+    let results_dir = builder.build.out.join("rustc-perf");
```

Comment:
Q: Can we benchmark cross target rustc? If we can, we should put this under the `build/$target-triple` directory. If we can't benchmark cross target rustc, then it's probably better to have this in `build/tmp` rather than having it in the build root.

Reply:
I have never tried that, but in theory it could work, because we don't execute the benchmarks. I will put it into `tmp` for now, it is a temporary directory after all.